lib/main.js

```diff
@@ -0,0 +1,290 @@
+    ' listeners, currently at ' + listenerCount);
+}
+
+// # EventEmitter
+// Constructor function, a replacement for ``EventEmitter`` that also allows
+// for asynchronous event listeners and allows event listeners to cancel events
+// if they return `false`. Event listeners may be either synchronous like DOM
+// events or asynchronous to better fit a async/await Promise style.  (Because
+// of that, making an event placed inside of a high performance loop is
+// probably a bad idea because the event listeners can introduce an
+// unpredictable delay.) The ``EventEmitter`` constructor function attaches
+// three "please don't touch" properties to itself (or its inheriting object)
+// to keep track of listeners and the maximum number of allowed listeners.
+export function EventEmitter() {
+  this._eventListeners = new Map();
+  this._onceListeners = new Map();
+  this._maxListeners = 10;
+  this.on('maxListenersPassed', printMaxListenersWarning.bind(this));
+}
+
+// ## emitSync
+// Operates just like ``emit`` except all listeners must be synchronous.
+EventEmitter.prototype.emitSync = function emitSync(eventName, ...args) {
+  // Short-circuit for emit calls with no listeners
+  if(!this._onceListeners.has(eventName) &&
+    !this._eventListeners.has(eventName)) {
+    return;
+  }
+
+  // Copy listeners arrays to ensure listener call order
+  const onceListeners = [...(this._onceListeners.get(eventName) || [])];
+  const eventListeners = [...(this._eventListeners.get(eventName) || [])];
+
+  // process once listeners first
+  if(onceListeners.length > 0) {
+    for(const listener of onceListeners) {
+      // remove from emitter list
+      // get again in case already removed
+      const _onceListeners = this._onceListeners.get(eventName);
+      if(_onceListeners) {
+        const idx = _onceListeners.indexOf(listener);
+        if(idx !== -1) {
+          _onceListeners.splice(idx, 1);
+        }
+      }
+      // emit 'removeListener'
+      this.emitSync('removeListener', eventName, listener);
+
+      // invoke listener, exceptions are not caught
+      const result = listener(...args);
+      // check for async result
+      if(isPromise(result)) {
+        throw new TypeError(`async listeners not allowed for "${eventName}"`);
+      }
+      // check if event canceled
+      if(result === false) {
+        return false;
+      }
+    }
+    // remove once list if empty
+    const _onceListeners = this._onceListeners.get(eventName);
+    if(_onceListeners.length === 0) {
+      this._onceListeners.delete(eventName);
+    }
+  }
+
+  // process regular listeners
+  for(const listener of eventListeners) {
+    // invoke listener, exceptions are not caught
+    const result = listener(...args);
+    // check for async result
+    if(isPromise(result)) {
+      throw new TypeError(`async listeners not allowed for "${eventName}"`);
+    }
+    // check if event canceled
+    if(result === false) {
+      return false;
+    }
+  }
+};
```

Comment:
It appears this returns `undefined` on success and `false` if canceled?  Is that proper?

Reply:
It's intended.  I'm not sure if it's proper. ;-)  It's basically what the API was doing before.  I was debating if we should really break the API and return something verbose like `{cancelled: true/false}` instead. This false/undefined thing is hard to understand.

Reply:
Tracking for future release: https://github.com/digitalbazaar/async-node-events/issues/5